Mathlib/MeasureTheory/Function/SimpleFunc.lean

```diff
@@ -511,48 +511,141 @@ theorem zpow_apply [DivInvMonoid β] (z : ℤ) (f : α →ₛ β) (a : α) : (f
-  Function.Injective.group (fun f => show α → β from f) coe_injective coe_one coe_mul coe_inv
-    coe_div coe_pow coe_zpow
+  fast_instance% Function.Injective.group (fun f => show α → β from f) coe_injective coe_one
+    coe_mul coe_inv coe_div coe_pow coe_zpow
 
 @[to_additive existing]
 instance instCommGroup [CommGroup β] : CommGroup (α →ₛ β) :=
-  Function.Injective.commGroup (fun f => show α → β from f) coe_injective coe_one coe_mul coe_inv
-    coe_div coe_pow coe_zpow
+  fast_instance% Function.Injective.commGroup (fun f => show α → β from f) coe_injective coe_one
+    coe_mul coe_inv coe_div coe_pow coe_zpow
+
+instance [Monoid K] [MulAction K β] : MulAction K (α →ₛ β) :=
+  fast_instance% Function.Injective.mulAction (fun f => show α → β from f) coe_injective coe_smul
 
 instance instModule [Semiring K] [AddCommMonoid β] [Module K β] : Module K (α →ₛ β) :=
-  Function.Injective.module K ⟨⟨fun f => show α → β from f, coe_zero⟩, coe_add⟩
+  fast_instance% Function.Injective.module K ⟨⟨fun f => show α → β from f, coe_zero⟩, coe_add⟩
     coe_injective coe_smul
 
 theorem smul_eq_map [SMul K β] (k : K) (f : α →ₛ β) : k • f = f.map (k • ·) :=
   rfl
 
+lemma smul_const [SMul K β] (k : K) (b : β) :
+    (k • const α b : α →ₛ β) = const α (k • b) := ext fun _ ↦ rfl
+
+lemma one_eq_const_one [One β] : (1 : α →ₛ β) = const α 1 := rfl
+
+instance [NonUnitalNonAssocSemiring β] : NonUnitalNonAssocSemiring (α →ₛ β) :=
+  fast_instance% Function.Injective.nonUnitalNonAssocSemiring (fun f => show α → β from f)
+    coe_injective coe_zero coe_add coe_mul coe_smul
+
+instance [NonUnitalSemiring β] : NonUnitalSemiring (α →ₛ β) :=
+  fast_instance% Function.Injective.nonUnitalSemiring (fun f => show α → β from f)
+    SimpleFunc.coe_injective coe_zero coe_add coe_mul coe_smul
+
+instance [NatCast β] : NatCast (α →ₛ β) where
+  natCast n := const _ (NatCast.natCast n)
+
+@[simp, norm_cast]
+lemma coe_natCast [NatCast β] (n : ℕ) :
+    ⇑(↑n : α →ₛ β) = fun _ ↦ ↑n := rfl
+
+instance [NonAssocSemiring β] : NonAssocSemiring (α →ₛ β) :=
+  fast_instance% Function.Injective.nonAssocSemiring (fun f => show α → β from f)
+    SimpleFunc.coe_injective coe_zero coe_one coe_add coe_mul coe_smul coe_natCast
+
+instance [NonAssocRing β] : NonAssocRing (α →ₛ β) where
+
+instance [NonUnitalCommSemiring β] : NonUnitalCommSemiring (α →ₛ β) :=
+  fast_instance% Function.Injective.nonUnitalCommSemiring (fun f => show α → β from f)
+    coe_injective coe_zero coe_add coe_mul coe_smul
+
+instance [CommSemiring β] : CommSemiring (α →ₛ β) where
+
+instance [NonUnitalCommRing β] : NonUnitalCommRing (α →ₛ β) :=
+  fast_instance% Function.Injective.nonUnitalCommRing (fun f => show α → β from f)
+    coe_injective coe_zero coe_add coe_mul coe_neg coe_sub coe_smul coe_smul
+
+instance [CommRing β] : CommRing (α →ₛ β) where
+
+instance [Semiring β] : Semiring (α →ₛ β) where
+
+instance [NonUnitalRing β] : NonUnitalRing (α →ₛ β) where
+
+instance [Ring β] : Ring (α →ₛ β) where
```

Comment:
We want these to follow the pattern too, I think.

Reply:
Wow...I'm going to have to pick your brain about this...